File: daily_problems/2024/10/1001/personal_submission/cf1677c_yawn_sean.py
# Submission link: https://codeforces.com/contest/1677/submission/283833723
def main():
    t = II()
    outs = []

    for _ in range(t):
        n = II()
        perm1 = LGMI()
        perm2 = LGMI()
        
        perm = [-1] * n
        for i in range(n):
            perm[perm1[i]] = perm2[i]
        
        k = 0
        vis = [0] * n
        
        for i in range(n):
            if not vis[i]:
                cnt = 0
                while not vis[i]:
                    vis[i] = 1
                    cnt += 1
                    i = perm[i]
                k += cnt // 2
        
        outs.append(2 * k * (n - k))

    print('\n'.join(map(str, outs)))

File: daily_problems/2024/10/1001/personal_submission/cf1677c_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include <bits/extc++.h>
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int t;
    cin >> t;

    while (t --) {
        int n;
        cin >> n;

        vector<int> perm1(n), perm(n);
        for (auto &v: perm1)
            cin >> v, v --;
        
        for (auto &v: perm1)
            cin >> perm[v], perm[v] --;
        
        vector<int> vis(n, 0);
        int k = 0;

        for (int i = 0; i < n; i ++) {
            if (!vis[i]) {
                int cnt = 0;
                while (!vis[i]) {
                    vis[i] = 1;
                    i = perm[i];
                    cnt ++;
                }
                k += cnt / 2;
            }
        }
        cout << 2ll * k * (n - k) << '\n';
    }

    return 0;
}